### DISCUSSION TITLE
Livewire has trouble managing alpine state / interactions on items rendered in loops

### DISCUSSION THREAD

**Discussion by edvordo** [livewire/livewire#7937] [7. Bugs (archived)]
### Livewire version

v3.4.4

### Laravel version

v10.43

### Which PHP version are you using?

PHP 8.2

### Steps To Reproduce

A fairly simple component with a list of items, that have an id, name and a category to filter the items down.

```php
<?php

namespace App\Livewire;

use Livewire\Component;

class Simple extends Component
{
    private array $list = [
        ['id' => '1', 'name' => 'item #01', 'category' => '1'],
        ['id' => '2', 'name' => 'item #02', 'category' => '1'],
        ['id' => '3', 'name' => 'item #03', 'category' => '1'],
        ['id' => '4', 'name' => 'item #04', 'category' => '1'],
        ['id' => '5', 'name' => 'item #05', 'category' => '1'],
        ['id' => '6', 'name' => 'item #06', 'category' => '1'],
        ['id' => '7', 'name' => 'item #07', 'category' => '1'],
        ['id' => '8', 'name' => 'item #08', 'category' => '1'],
        ['id' => '9', 'name' => 'item #09', 'category' => '1'],
        ['id' => '10', 'name' => 'item #10', 'category' => '2'],
        ['id' => '11', 'name' => 'item #11', 'category' => '2'],
        ['id' => '12', 'name' => 'item #12', 'category' => '2'],
        ['id' => '13', 'name' => 'item #13', 'category' => '2'],
        ['id' => '14', 'name' => 'item #14', 'category' => '2'],
        ['id' => '15', 'name' => 'item #15', 'category' => '2'],
        ['id' => '16', 'name' => 'item #16', 'category' => '2'],
        ['id' => '17', 'name' => 'item #17', 'category' => '3'],
        ['id' => '18', 'name' => 'item #18', 'category' => '3'],
        ['id' => '19', 'name' => 'item #19', 'category' => '3'],
        ['id' => '20', 'name' => 'item #20', 'category' => '3'],
        ['id' => '21', 'name' => 'item #21', 'category' => '3'],
        ['id' => '22', 'name' => 'item #22', 'category' => '3'],
        ['id' => '23', 'name' => 'item #23', 'category' => '3'],
        ['id' => '24', 'name' => 'item #24', 'category' => '3'],
        ['id' => '25', 'name' => 'item #25', 'category' => '3'],
        ['id' => '26', 'name' => 'item #26', 'category' => '3'],
        ['id' => '27', 'name' => 'item #27', 'category' => '3'],
        ['id' => '28', 'name' => 'item #28', 'category' => '3'],
        ['id' => '29', 'name' => 'item #29', 'category' => '3'],
        ['id' => '30', 'name' => 'item #30', 'category' => '3'],
    ];

    public string $category = '';
    
    public function getItemsProperty()
    {
        return collect($this->list)
            ->when(
                false === empty($this->category),
                fn ($items) => $items->filter(fn ($item) => $item['category'] === $this->category)
            )
            ->values();
    }
    
    public function render()
    {
        return view('livewire.simple');
    }
}
```
The associated blade file, where a portion of the Livewire component is an alpine component, that keeps track of currently mouseover-ed item and selected item, both represented on the list with a different background.
```blade
<div class="flex w-full h-full items-center justify-center text-gray-300 px-16">
    <div>Filtered count: {{ count($this->items) }}</div>

    <div x-data="{
        activeId: '', 
        selectedId: '', 
        setActive(id) { 
            this.activeId = id; 
        }, 
        setSelected(id) { 
            this.selectedId = id; 
        }
    }">
        <div class="flex flex-col">
            <label>Category</label>
            <input type="text" wire:model.live="category" />
        </div>

        <div class="flex flex-wrap">
            @forelse($this->items as $item)
                <div
                    wire:key="item-{{ $item['id'] }}"
                    :class="{
                        'bg-gray-700': activeId === '{{ $item['id'] }}',
                        'bg-green-700': selectedId === '{{ $item['id'] }}'
                    }"
                    class="p-1 border border-gray-500 text-xs w-1/4"
                    x-on:click="setSelected('{{ $item['id'] }}');"
                    x-on:mouseover="setActive('{{ $item['id'] }}');"
                >
                    <span>{{ $item['name'] }}</span>
                    <span>{{ $item['category'] }}</span>
                </div>
            @empty
                <div
                    wire:key="item-empty"
                    class="p-2 border border-gray-500 w-1/4"
                    x-on:click="setSelected('empty');"
                    x-on:mouseover="setActive('empty');"
                >
                    <span>N/A</span>
                    <span>N/A</span>
                </div>
            @endforelse
        </div>
    </div>
</div>
```
Now type any of the categories into the input, lets say `3`, the list gets filtered down and everything still works properly.
But remove that value from the input and some of the previously filtered items no longer work, as in no cell reacts to either a `x-on:mouseover` or a `x-on:click` events.

### Repository or Wirebox URL

https://wirebox.app/b/4jdrv

### Please confirm (incomplete submissions will not be addressed)

- [X] I have provided easy and step-by-step instructions to reproduce the bug.
- [X] I have provided code samples as text and NOT images.
- [X] I understand my bug report will be removed if I haven't met the criteria above.

**Comment by MrCrayon** [comment]
I can't see anything wrong in your code so it looks like a bug to me.
As a work around you can set `wire:key="item-{{ $category.'-'.$item['id'] }}"`

**Reply by edvordo** [reply]
Thank you for validating my code, that helps!

Well, my actual component I based this issue on is quite a bit more complicated, but yea, I forced a workaround in similar spirit by adding `microtime(true)` to the `wire:key` attribute, which has fixed the issue for the time being. But it's ugly and a proper fix from livewire / alpine side would be welcome.

**Comment by Ing-LuisGuerrero** [comment]
I also encountered this error. It seems that Alpine components are sometimes not initialized. Using the mentioned workaround feels like it undermines the purpose of using wire:key, as we force the element to re-render.

Here's my [wirebox](https://wirebox.app/b/x9oep):
1. Search for "Todo 4". It will render "Todo 4" and "Todo 44"
2. Delete the input value by searching again.
3. Click the options button on Todo 44; the dropdown will not appear.

Here's the Alpine data for "Todo 44" from the first and last renderings:
<img width="811" alt="imagen" src="https://github.com/livewire/livewire/assets/48639261/353127b5-166b-4731-99ba-2e6a9d6dca1e">

I kindly hope this could be addressed soon.

**Comment by EdgarsJoja** [comment]
I think I have encountered this issue  or similar as well. I observed, that on re-render in some specific scenarios items drop events (same as you mentioned), e.g. `click`. From what I understand this most likely would be due to morphing issues and the usual solution supposedly is to make sure that `wire:key` is correctly set. But even after adding that, the issue did not disappear.
What I noticed is that changing order of the rendered items seem to fix particular scenario (again, seems to indicate morphing cause, from what I understand), but that is obviously no real fix.
As of right now, I haven't found a solution, but because it happened in a very specific scenario for me, I pushed it further down the to-do list.

**Comment by SEUH** [comment]
Here's another example of this. Not really sure why not more people are having issues with this. Adding to an array at the end works just fine, but adding a value befor existing objects completely breaks state. Press reverse twice also breaks state. There must be a bug in the morphing code. Setting `inject_morph_markers` to `false` has no effect (in my testing i actually had more problems with setting that to false...).

EDIT: this has already been reported 4 months ago. https://github.com/livewire/livewire/discussions/7785. Since Caleb said this was fixed in https://github.com/livewire/livewire/pull/7850 i checked my alpine version and the frontend uses `v3.13.8`, so i guess it's broken again.

EDIT **IMPORTANT**: There is a [test case](https://github.com/livewire/livewire/blob/main/src/Tests/AlpineMorphingBrowserTest.php#L90) that "covers" this bug but the fun part is that this bug doesn't apper with only 3 items in the array, you need at least 4 :D

```php
<?php

namespace App\Livewire;

use Livewire\Component;

class ForeachSameElementsDifferentOrderBug extends Component
{
    public $changedOrder = false;

    public $rawData;

    public function __construct()
    {
        $this->rawData = [
            (object)['id' => 1, 'text' => 1],
            (object)['id' => 2, 'text' => 2],
            (object)['id' => 3, 'text' => 3],
            (object)['id' => 4, 'text' => 4],
            (object)['id' => 5, 'text' => 5],
            (object)['id' => 6, 'text' => 6],
            (object)['id' => 7, 'text' => 7],
            (object)['id' => 8, 'text' => 8],
            (object)['id' => 9, 'text' => 9],
            (object)['id' => 10, 'text' => 10],
        ];
    }

    public function toggleOrder()
    {
        $this->changedOrder = !$this->changedOrder;
    }

    public function getDataProperty()
    {
        return $this->changedOrder ? array_reverse($this->rawData) : $this->rawData;
    }

    public function increaseCounter()
    {
        $this->counter++;
    }

    public function addItemStart()
    {
        $idx = count($this->rawData) + 1;
        array_unshift($this->rawData, (object)['id' => $idx, 'text' => $idx]);
    }

    public function addItemEnd()
    {
        $idx = count($this->rawData) + 1;
        array_push($this->rawData, (object)['id' => $idx, 'text' => $idx]);
    }

    public function render()
    {
        return view('livewire.foreach-same-elements-different-order-bug');
    }
}
```

```blade
<div>
    <div>
        <button wire:click="addItemEnd">add item at end (works)</button>
        <button wire:click="addItemStart">add item at start (broken)</button>
        <button wire:click="toggleOrder">reverse list</button>
    </div>
    <hr>
    <div>
        @foreach($this->data as $row)
            <div wire:key="{{ $row->id }}" style="display: flex">
                <div>{{ $row->id }}</div>
                <div style="display: flex">
                    {{ $row->text }}
                    <div x-data="{myOwnValue: true}">
                        <div x-show="myOwnValue">should be visible</div>
                    </div>
                </div>
            </div>
        @endforeach
    </div>
</div>
```

**Comment by aintyourcat** [comment]
I also faced a similar issue, i explained it thoroughly here: https://github.com/livewire/livewire/discussions/8841
